contracts/cw20-atomic-swap/src/contract.rs

```diff
@@ -31,22 +37,48 @@ pub fn handle<S: Storage, A: Api, Q: Querier>(
     msg: HandleMsg,
 ) -> StdResult<HandleResponse> {
     match msg {
-        HandleMsg::Create(msg) => try_create(deps, env, msg),
+        HandleMsg::Create(msg) => try_create(
+            deps,
+            env.clone(),
+            msg,
+            Balance::Native(env.message.sent_funds),
+        ),
         HandleMsg::Release { id, preimage } => try_release(deps, env, id, preimage),
         HandleMsg::Refund { id } => try_refund(deps, env, id),
+        HandleMsg::Receive(msg) => try_receive(deps, env, msg),
+    }
+}
+
+pub fn try_receive<S: Storage, A: Api, Q: Querier>(
+    deps: &mut Extern<S, A, Q>,
+    env: Env,
+    wrapper: Cw20ReceiveMsg,
+) -> StdResult<HandleResponse> {
+    let msg: ReceiveMsg = match wrapper.msg {
+        Some(bin) => from_binary(&bin),
+        None => Err(StdError::parse_err("ReceiveMsg", "no data")),
+    }?;
+    let token = Cw20Coin {
+        address: deps.api.canonical_address(&env.message.sender)?,
+        amount: wrapper.amount,
+    };
+    match msg {
+        ReceiveMsg::Create(create) => try_create(deps, env, create, Balance::Cw20(token)),
     }
 }
 
 pub fn try_create<S: Storage, A: Api, Q: Querier>(
     deps: &mut Extern<S, A, Q>,
     env: Env,
     msg: CreateMsg,
+    balance: Balance,
 ) -> StdResult<HandleResponse> {
     if !is_valid_name(&msg.id) {
         return Err(StdError::generic_err("Invalid atomic swap id"));
     }
 
-    if env.message.sent_funds.is_empty() {
+    // FIXME? What about a non-empty array, but with zero-valued coins?
```

Comment:
That should be rejected. We can have a "normalize" function that removes all 0 values

Reply:
OK. I'll change the comment now, and address that as part of #73.